### PR TITLE
Sort individuals

### LINKED ISSUE #1197: sort should ensure ordering of individual parents (also, mutation parents?)
`sort( )` should produce valid tables; with #1192 one of the requirements not ensured by `sort` is that parent individuals come before children. We should do this! Also, I think we should do the same for mutations. (I'm only proposing sorting so that mutations who have their parent mutations specified get sorted correctly; otherwise we'd have to compute the mutation parents.)

Since sort is a bottleneck in forwards sims, and this sortedness is not required by simplify, perhaps this should be optional. (although it sounds annoying to make it optional)

### PR DESCRIPTION
**Pull request by benjeffery**
Sort individuals. Fixes #1197 #1138.

WIP!

Currently only a python implementation (deliberately written in a style to aid translation to C). Needs more tests etc, but thought it worth sharing progress.

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Looks very nice, eminently C-able!

**Comment by benjeffery** [comment]
Was hoping to have this done yesterday, but taking longer than expected - the initial implementation's sort wasn't stable, that's fixed now. Currently fixing up tests in python side.

**Comment by jeromekelleher** [comment]
> the initial implementation's sort wasn't stable, 

It's the gift that keeps on giving! Well done for spotting early.

**Comment by petrelharp** [comment]
Nicely done. So the sort here is decreasing by "minimum number of hops to a childless descendant, and then smallest position in the table among the closest childless descendant".

**Comment by benjeffery** [comment]
@petrelharp Yes! That's a great way of stating it. Should have this polished off today.

**Comment by benjeffery** [comment]
While working on the python tests I realised we also need to tweak the canonicalisation as the sort there no longer meets the requirements of a tree sequence.

**Comment by benjeffery** [comment]
We also have the classic msprime chicken and egg as all the sort testing code uses msprime that doesn't store the pedigree. Will add some `wf_sim` tree sequences to the sort tests for now.

**Comment by petrelharp** [comment]
> While working on the python tests I realised we also need to tweak the canonicalisation as the sort there no longer meets the requirements of a tree sequence.

But `canonicalize` just uses the standard sort [with an enhanced mutation sort](https://github.com/tskit-dev/tskit/blob/main/c/tskit/tables.c#L9058) so it'll get this change?

**Comment by benjeffery** [comment]
> > While working on the python tests I realised we also need to tweak the canonicalisation as the sort there no longer meets the requirements of a tree sequence.
> 
> But `canonicalize` just uses the standard sort [with an enhanced mutation sort](https://github.com/tskit-dev/tskit/blob/main/c/tskit/tables.c#L9058) so it'll get this change?

Ah, so I think the sort of individuals will be valid then. However, their sort order will depend on the ordering of the parents in each individual though - is that ok for canonicalisation or do we need to have a canonical ordering of the parent arrays?

**Comment by petrelharp** [comment]
> Ah, so I think the sort of individuals will be valid then. However, their sort order will depend on the ordering of the parents in each individual though - is that ok for canonicalisation or do we need to have a canonical ordering of the parent arrays?

Oh, right - canonicalize uses `subset` to put the parents in order by node. That happens [before sort](https://github.com/tskit-dev/tskit/blob/main/c/tskit/tables.c#L9068), so I guess that with this change, canonicalise will sort individuals by "minimum number of hops to a childless descendant, and then smallest **node** among the closest childless descendant". It's still canonical, given the node ordering, I think? This will need some modifications to `canonicalise`'s docs, anyhow.

I guess that now `subset` can produce invalid tree sequences, though, since it reorders individuals by their nodes.

**Comment by jeromekelleher** [comment]
> I guess that now subset can produce invalid tree sequences, though, since it reorders individuals by their nodes.

Ah, that's unfortunate...

**Comment by benjeffery** [comment]
> I guess that now `subset` can produce invalid tree sequences, though, since it reorders individuals by their nodes.
Yes, this is true, I guess it can just sort the individuals as a last step? 

It doesn't seem like `subset` reorders the parent array here, so the final sort order is still dependant on the order the parents were specified in.

Also I can't see where `subset` remaps the ids in `individual.parent`?

**Comment by petrelharp** [comment]
Ok, how about this: 
- have `subset` leave the individual table ordering alone (so, still subset individuals, but don't re-order them), and 
- come up with a canonical ordering for individuals, that either happens here, or if it's difficult, that `canonicalise` does.

**Comment by petrelharp** [comment]
As for the order: modifying what you've got here it'd be easy to efficiently compute the total number of descendants of each individual (by recursively adding up the values of all children), so then we could have (num_descendants, smallest referring node) as a canonical sort key. I dunno whether to swap that in here or just do it in canonicalize - what you have here is nice because it doesn't use `qsort`, but then why would we bother implementing two different sorts in different places?

**Comment by jeromekelleher** [comment]
> I dunno whether to swap that in here or just do it in canonicalize - what you have here is nice because it doesn't use qsort, but then why would we bother implementing two different sorts in different places?

Well, one argument would be one is O(n) and the other is O(n log n). I don't have a strong opinion about it though.

**Comment by benjeffery** [comment]
As another option, I think if `subset` also sorts the parents in `individual.parents` then applying this sort will be canonical. 

I don't have enough information to know how to prioritise the speed of sort vs the complexity of having two sorts. Sorting is a very common operation though so I err to having that be O(n).

If it makes sense I'd like to finish this PR as-is and then fix the canonical issues in a separate one?

**Comment by petrelharp** [comment]
>  I think if subset also sorts the parents in individual.parents then applying this sort will be canonical.

Um, I guess this is true? I'm a little unsure about this, given that the outcome of this sort depends on the order of parents. 

> If it makes sense I'd like to finish this PR as-is and then fix the canonical issues in a separate one?

That makes sense, but I think we should make sure we have good path forward: we'd like to at least make sure that if you canonicalise, a subsequent sort doesn't change anything, for instance.  And, that the order induced by canonicalise does not depend on the original order of the parent table, just on the order of the node table. And probably also not on the order that the different parents of a given individual are specified.

I'll think some more about this (although I'm ready to be done with sorting... =)

**Comment by jeromekelleher** [comment]
Agree with all the above, but:

> : we'd like to at least make sure that if you canonicalise, a subsequent sort doesn't change anything, for instance. 

I wouldn't get too hung up on this; once the output of sort is guaranteed to be legal input for tables that's all that really matters. Canonical must be canonical though, there can be only one output from ``canonicalise``.

**Comment by petrelharp** [comment]
> I wouldn't get too hung up on this; once the output of sort is guaranteed to be legal input for tree sequences that's all that really matters. Canonical must be canonical though, there can be only one output from canonicalise.

Ok, wise words there.  Here's my proposal, then:
1. make it so `subset` doesn't reorder the individual table (this is easy and makes it easier to understand anyhow)
2. let `sort` do whatever (so: this PR is fine)
3. change `canonicalize` so that it sorts individuals on `(num descendants, id of first referring node)` - this will be a bit annoying but straightforward

I've given it a fair bit of thought and come up with no better solution.

**Comment by jeromekelleher** [comment]
Sounds like a plan, thanks @petrelharp.

So, I guess we merge the sort stuff here and file some issues to track the other two (which we resolve before release)?

**Comment by petrelharp** [comment]
I've made a good start at `subset` and `canonicalise` (in #1226), and turned up another issue (`union`); I think I'll wait until this is merged to finish since I'm starting to step on the toes of this PR.

**Comment by benjeffery** [comment]
Needs a squash and a changelog, but I finally think this is ready for review. I had to do the subset individual re-mapping (#1223 ) to get the tests to pass.

File: c/tskit/tables.c
Comment (petrelharp) [inline c/tskit/tables.c:4964]: you don't use `individual` and `copy_individual` at the same time, so maybe they should both be `individual`?

File: python/tests/tsutil.py
Comment (petrelharp) [inline python/tests/tsutil.py:1105]: did you mean to put the migration sort stuff in here?

**Review by petrelharp — APPROVED** [review]
Looks good! Very small comments, and: how about throwing in a test that inserts some `NULL` parents in addition to the actual parents? (Or, maybe we should just disallow them.)

File: python/tests/tsutil.py
Comment (benjeffery) [reply python/tests/tsutil.py:1105]: Yes, otherwise a comparison of the python and c sorting fails.

File: c/tskit/tables.c
Comment (jeromekelleher) [inline c/tskit/tables.c:4996]: nit - might as well use ``++`` here for consistency

File: c/tskit/tables.c
Comment (jeromekelleher) [inline c/tskit/tables.c:5041]: I agree with Peter - ``copy_individual`` doesn't really add any clarity over ``individual`` here.

File: c/tskit/tables.c
Comment (jeromekelleher) [inline c/tskit/tables.c:5059]: Might be worth taking a reference to the node table above (like we do for the individual table) to avoid a bit of pointer chasing in this loop.

File: c/tskit/tables.c
Comment (jeromekelleher) [inline c/tskit/tables.c:5145]: I know it's redundant, but let's stick the ``if (ret != 0)`` stanza in here. It's easy to miss in diffs if someone updates the code subsequently, and forgets to put it in.

File: c/tskit/tables.c
Comment (jeromekelleher) [inline c/tskit/tables.c:9882]: We should try to cover this in tests? (or is that coming in one of the follow-ups?)

File: python/tests/tsutil.py
Comment (jeromekelleher) [inline python/tests/tsutil.py:1006]: It doesn't really matter I'd imagine, but 
```
migration_i = tables.migrations[i]
migration_j = tables.migrations[j]
ret = migration_i.time - migration_j.time
# etc
```
would avoid repeatedly building and returning numpy arrays for each of the columns.

File: python/tests/tsutil.py
Comment (jeromekelleher) [inline python/tests/tsutil.py:1686]: typo, indivial

**Review by jeromekelleher — APPROVED** [review]
Very nice, a few minor comments. The algorithm is quite beautiful!

**Comment by jeromekelleher** [comment]
> Looks good! Very small comments, and: how about throwing in a test that inserts some NULL parents in addition to the actual parents? (Or, maybe we should just disallow them.)

I think we should allow NULL parents - there will be times when having parents ``[-1, -1]`` will be handy and others when parents=``[]`` is more useful. We should allow both, I think.

File: c/tskit/tables.c
Comment (petrelharp) [reply c/tskit/tables.c:9882]: yes, this is in #1226.

File: c/tskit/tables.c
Comment (benjeffery) [reply c/tskit/tables.c:5142]: Done

File: c/tskit/tables.c
Comment (benjeffery) [reply c/tskit/tables.c]: Done

File: c/tskit/tables.c
Comment (benjeffery) [reply c/tskit/tables.c]: Done

File: c/tskit/tables.c
Comment (benjeffery) [reply c/tskit/tables.c]: Done

File: python/tests/tsutil.py
Comment (benjeffery) [reply python/tests/tsutil.py]: Done

**Pull request merged**
